virttest/utils_net.py

```diff
@@ -4080,35 +4085,40 @@ def get_default_gateway_json(
         raise ValueError(f"Unrecognized IP version {ip_ver}")
     run_func = session.cmd_output if session else process.getoutput
 
-    try:
-        ip_output_str = run_func(cmd).strip()
-        LOG.debug(f"ip route output:\n{ip_output_str}")
-        ip_route = json.loads(ip_output_str)
-    except Exception as why:
-        LOG.error(f'Failed to get output of "{cmd}" command. Reason: {str(why)}')
-        return None
-
-    default_route_list = [x for x in ip_route if x["dst"] == "default"]
-    if force_dhcp:
-        default_route_list = [
-            x for x in default_route_list if x.get("protocol") == "dhcp"
-        ]
-    if target_iface:
-        LOG.debug(f"Get default gateway only for: {target_iface}")
-        non_multi = [x for x in default_route_list if x.get("dev") == target_iface]
-        multi = [
-            p
-            for x in default_route_list
-            if "nexthops" in x
-            for p in x["nexthops"]
-            if p.get("dev") == target_iface
-        ]
-        default_route_list = non_multi + multi
-    LOG.debug(f"default_route_list: {default_route_list}")
+    def _get_default_route_list():
+        try:
+            ip_output_str = run_func(cmd).strip()
+            LOG.debug(f"ip route output:\n{ip_output_str}")
+            ip_route = json.loads(ip_output_str)
+        except Exception as why:
+            LOG.error(f'Failed to get output of "{cmd}" command. Reason: {str(why)}')
+            return None
 
-    if len(default_route_list) == 0:
+        nonlocal default_route_list
+        default_route_list = [x for x in ip_route if x["dst"] == "default"]
+        if force_dhcp:
+            default_route_list = [
+                x for x in default_route_list if x.get("protocol") == "dhcp"
+            ]
+        if target_iface:
+            LOG.debug(f"Get default gateway only for: {target_iface}")
+            non_multi = [x for x in default_route_list if x.get("dev") == target_iface]
+            multi = [
+                p
+                for x in default_route_list
+                if "nexthops" in x
+                for p in x["nexthops"]
+                if p.get("dev") == target_iface
+            ]
+            default_route_list = non_multi + multi
+        LOG.debug(f"default_route_list: {default_route_list}")
```

Comment:
IIUC it's the first time we observe this so having to retry shouldn't be necessary but important here to stabilize our test runs. Therefore, would it make sense to LOG.warning or something similar whenever there wasn't a default route?

Reply:
Updated